javaslang/src/main/java/javaslang/collection/Vector.java

```diff
@@ -51,12 +53,48 @@
     final T[] leading, trailing;
 
     private Vector(T[] leading, VectorTree<T> middle, T[] trailing) {
+        /* Normalize the 3 containers */
+        if (isNullOrEmpty(leading)) {
```

Comment:
these assure the invariants presented in the description, i.e.
* that the `leading` is only empty if the whole vector is empty;
* the `middle` tree's leafs have size equal to the `branchingFactor()`;
* the `trailing` is only empty, if the `midle` tree is also.